### ISSUE TITLE
@custom-media directive fails

### ISSUE THREAD

**Issue by josephfinlayson** [less/less.js#2639]
Input LESS

```
.flex-grid {
    @import  "../../../../../bower_components/flexboxgrid/src/css/flexboxgrid.css";
}
```

Output CSS

```
.flex-grid {
  @import "../../../../bower_components/flexboxgrid/src/css/flexboxgrid.css";
}
```

How do I namespace CSS imports? Casting this file to less doesn't work as it uses newer CSS features. e.g.: `@custom-media --md-viewport only screen and (min-width: 62em);`

**Comment by seven-phases-max** [comment]
If you don't want CSS to be parsed use [`inline`](). Though for most of the namespaced imports only `less` import can valid since the compiler must parse and understand content to actually apply the namespacing properly, hence Less has to support `@custom-media` natively to parse and understand it too.

You can make a feature-request to support `@custom-media` natively, though adding drafted features usually a bit annoying since those tend to change several times before stabilized

**Comment by josephfinlayson** [comment]
I see, I thought it was already part of the spec. I guess this would be part of a wider discussion on when LESS should support newer CSS features -  custom media is part of [this](http://drafts.csswg.org/mediaqueries-4/) draft spec.

**Comment by matthew-dean** [comment]
IMO Less will support features when they are implemented in more than one major browser, which I believe is along the same guidelines the W3C uses from moving something from Candidate Recommendation to Recommended status.

However, this is a gray area, because technically, in CSS, you can write any @-rule you want, and if the browser doesn't support it, it simply ignores it. Less doesn't really currently lint properties to test if they're valid, so explicitly passing only certain at rules is a little inconsistent but probably pragmatically necessary. Ideally though, if Less was going to act as a linter, it would simply check to see that CSS statements were correctly constructed, and that Less statements were valid.

So, I would support simply allowing any @-rule that passes the CSS syntax rules for valid construction of an @-rule, just like we do with blocks and parameters / values.

**Comment by seven-phases-max** [comment]
>So, I would support simply allowing any @-rule that passes the CSS syntax rules for valid construction of an @-rule

This makes sense. Actually currently Less does pass any unknown at-rule followed by `{}` block. So `@custom-media` (and any other unknown directive ending with `;`) is a more rare exception. I suppose it's not that difficult to modify at-rule parser function to support either `{}` or `;`-ending directives, it's just a matter of inventing more tests and making sure that a more relaxed parsing won't hit performance too much (obviously because too relaxed `@*` interferes with variables syntax).

**Comment by stevenvachon** [comment]
Another reason why Less should be a PostCSS plugin.

**Comment by matthew-dean** [comment]
@stevenvachon Making Less a PostCSS plugin would not change anything except parsing. The rules would still need to be evaluated by Less.

**Comment by seven-phases-max** [comment]
An universal answer is: "So are you planning to write one?" ;)

**Comment by stevenvachon** [comment]
No time; I have other projects. Less shouldn't have plugins as we already have PostCSS.

**Comment by stevenvachon** [comment]
Perhaps we need a `literal()` for now so that we can still have future-CSS code.

**Comment by seven-phases-max** [comment]
>Perhaps we need a `literal()` for now so that we can still have future-CSS code.

There're already `@import (inline)`, `~""` and after all `e()`.

**Comment by matthew-dean** [comment]
> and it stays absolutely the same story if it's implemented as a PostCSS plugin.

Correct. This is what I was saying. PostCSS doesn't solve anything without a preprocessing plugin to evaluate it (or "transform it" as PostCSS calls it). Even the author of PostCSS specifies that all that PostCSS technically is just a parser which creates an AST.

That said, I'm "working on" a postcss-less plugin. I say it in quotes because it depends on having extra time.

**Comment by stevenvachon** [comment]
Once many CSS features are implemented in the browsers, you guys will have to implement them in Less. They're already available as plugins for PostCSS, so it'd be less work for you and less waiting for us in the future.

`@import (inline) "file.css";` worked well, by the wall. Thanks.

**Comment by seven-phases-max** [comment]
I think I'll merge this to #2770 (this one is earlier than but to not reassign labels).
In fact *if* #2770 is properly fixed, the example in the initial post would work as expected since unknown at-rules are marked for bubbling anyway (and this should be fine for most of unknown/custom at-rules).

**Comment by seven-phases-max** [comment]
Merged to #2770.

**Comment by seven-phases-max** [comment]
Re-opening this since #2770 is fixed, but bubbling for non-block at-rules is not enabled (see #2783).
